feat: Allow metadata filtering for fetch records endpoints

frascuchon opened the pull request:
<!-- Thanks for your contribution! As part of our Community Growers initiative 🌱, we're donating Justdiggit bunds in your name to reforest sub-Saharan Africa. To claim your Community Growers certificate, please contact David Berenstein in our Slack community or fill in this form https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

The purpose of this PR is to add support for metadata filtering to API endpoints fetching records:
- `/api/v1/datasets/:dataset_id/records`
- `/api/v1/me/datasets/:dataset_id/records`

by using the same metadata query pattern defined for the search endpoint 

To support this, the `SearchEngine` have been modify to allow searches without a query (just for filtering)

Refs #3748 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] New feature (non-breaking change which adds functionality)
- [x] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

Some local tests have been done.

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

frascuchon committed:
feat: `MetadataProperty` database model class

frascuchon committed:
feat: Define alembic migrations for `metadata_properties` table

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
chore: change naming to MetadataPropertyType

frascuchon committed:
fix: exposing server enums from server models module

frascuchon committed:
feat: define new API endpoint for creating and listing dataset metadata properties

frascuchon committed:
tests: adding new tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
chore: Update changelog

frascuchon committed:
Merge branch 'feat/list-and-creation-endpoints-for-dataset-metadata-properties' of github.com:argilla-io/argilla into feat/list-and-creation-endpoints-for-dataset-metadata-properties

frascuchon committed:
tests: Add missing tests for listing dataset metadata properties

frascuchon committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' into feat/list-and-creation-endpoints-for-dataset-metadata-properties

frascuchon committed:
feat(search): Metadata filtering support for `search` method

Also, brings all refactor already done in the similarity search feature branch.

This will help to merge changes after this feature is done.

frascuchon committed:
feat: define metadata filter schemas and change search endpoint

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
fix: wrapping metadata filter validations to return proper status code

frascuchon committed:
fix(search): applying range filter for 0.0 values

Also, the query param is now optional. Searches can be applied only for filtering data.

frascuchon committed:
tests: Adding new test cases.

frascuchon committed:
Merge branch 'feat/support-metadata-filtering-for-search-endpoint' of github.com:argilla-io/argilla into feat/support-metadata-filtering-for-search-endpoint

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' into feat/support-metadata-filtering-for-search-endpoint

frascuchon committed:
Merge branch 'feat/support-metadata-filtering-for-search-endpoint' of github.com:argilla-io/argilla into feat/support-metadata-filtering-for-search-endpoint

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
wip

frascuchon committed:
tests: Add missing tests

frascuchon committed:
feat(records): Allow filter by metadata for fetch records endpoints

frascuchon committed:
tests: Adding tests for listing records with metadata filters

frascuchon committed:
tests: Adding missing tests

frascuchon committed:
Merge branch 'feat/allow-metadata-filter-for-fetch-records-endpoints' of github.com:argilla-io/argilla into feat/allow-metadata-filter-for-fetch-records-endpoints

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' into feat/allow-metadata-filter-for-fetch-records-endpoints

gabrielmbmb commented on src/argilla/server/apis/v1/handlers/datasets.py:
```diff
@@ -332,6 +372,36 @@ async def create_dataset_metadata_property(
+    except ValueError as err:
+        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(err))
+
+
```
this is duplicate

frascuchon replied:
Done (again) 🙃

gabrielmbmb commented on src/argilla/server/apis/v1/handlers/datasets.py:
```diff
@@ -474,6 +542,47 @@ async def search_dataset_records(
 
+async def _build_metadata_filters(
+    db: "AsyncSession", dataset: Dataset, parsed_metadata: List[MetadataParsedQueryParam]
+):
```
```suggestion
) -> List["MetadataFilter"]:
```

gabrielmbmb replied:
will need to import `MetadataFilter`

frascuchon committed:
feat(engine): Allow passing `sort_by` configuration to `search` method (#3825)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR is part of the work in
https://github.com/argilla-io/argilla/pull/3819 and adds configurable
`sort_by` options for the `SearchEngine.search` method.

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing
functionality)
- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

Some local test have been done 

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

gabrielmbmb commented on src/argilla/server/apis/v1/handlers/datasets.py:
```diff
@@ -150,15 +160,33 @@ async def list_current_user_dataset_records(
+        record_ids = [response.record_id for response in search_responses.items]
+        records = await datasets.get_records_by_ids(
+            db=db, dataset_id=dataset_id, record_ids=record_ids, include=include
+        )
```
maybe we could move this to a function, so it could be reused in both endpoints

frascuchon replied:
Done

frascuchon committed:
fix(engine): filter by response status with no user info

frascuchon committed:
refactor: combine code in a single function

frascuchon committed:
tests: Add tests for filtering responses with no user info

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
fix: Remove duplicated code

frascuchon committed:
Merge branch 'feat/allow-metadata-filter-for-fetch-records-endpoints' of github.com:argilla-io/argilla into feat/allow-metadata-filter-for-fetch-records-endpoints

frascuchon committed:
chore: Apply suggestion

frascuchon committed:
chore: Change name for all status responses in elasticsearch

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci
